fix: Disallow changing username and email in case of Organization email invite

hariombalhara opened the pull request:
## What does this PR do?

Fixes following issues - Regression from https://github.com/calcom/cal.com/pull/11421/files#diff-e7c1df7ced03badb4ff3248d2f29752d2a9c968747c96b179aadc08e350d7a1eL166
- Member Invite: Invited member was able to change the email and username
- Invite Link: Person signing up using invite link was required to provide username and he could potentially provide a different username than  the one derived from email.

Also, adds more assertions and flows in existing signup tests
- Verifies that username is correctly derived from email when the email is of the org.
- Verifies that username is correctly derived from email when the email is not of the org.
- Verifies that non-org signup uses the provided username.

Implementation Details:
`username` validation and computation has been added at backend as well so that even if username is changed at backend, it would ignore it in case of organization invite.


Reported at https://threads.com/thread/34551320020/34551320023?s=f5bBRyt2WeiKaHf1zCFK57

Member Invite - Now:
![image](https://github.com/calcom/cal.com/assets/1780212/2214cdbd-d407-46c2-b25a-2c7ecd5c78f7)
Member Invite - Before:
![image](https://github.com/calcom/cal.com/assets/1780212/759bf9a4-4202-41ff-b6c7-a14bc24822c4)

Invite Link - Now
![image](https://github.com/calcom/cal.com/assets/1780212/e4ac6515-14d8-4f31-959f-43fe04464224)
Invite Link -Before
![image](https://github.com/calcom/cal.com/assets/1780212/e12f5827-832a-4f67-99db-249772af8ee9)


## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Tests (Unit/Integration/E2E or any other test)

## How should this be tested?
- Invite an email to an org say member1@gmail.com
- Go to the email received and reach the signup page from the link in the email
- Observe the username and email fields. They should be disabled.

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

hariombalhara committed:
Disallow changing username and email in case of Organization email invite

hariombalhara committed:
Disallow changing username and email in case of Organization email invite

hariombalhara commented:
Current dependencies on/for this PR:
* main
  * **PR #12735** <a href="https://app.graphite.dev/github/pr/calcom/cal.com/12735?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/calcom/cal.com/12735?utm_source=stack-comment).

hariombalhara commented on apps/web/pages/signup.tsx:
```diff
@@ -237,7 +237,7 @@ export default function Signup({
 
   return (
-    <div className="light bg-muted 2xl:bg-default flex min-h-screen w-full flex-col items-center justify-center [--cal-brand-emphasis:#101010] [--cal-brand:#111827] [--cal-brand-text:#FFFFFF] [--cal-brand-subtle:#9CA3AF] dark:[--cal-brand-emphasis:#e1e1e1] dark:[--cal-brand:white] dark:[--cal-brand-text:#000000]">
+    <div className="light bg-muted 2xl:bg-default flex min-h-screen w-full flex-col items-center justify-center [--cal-brand-emphasis:#101010] [--cal-brand-subtle:#9CA3AF] [--cal-brand-text:#FFFFFF] [--cal-brand:#111827] dark:[--cal-brand-emphasis:#e1e1e1] dark:[--cal-brand-text:#000000] dark:[--cal-brand:white]">
```
Classes re-ordering automatically done by eslint.

hariombalhara commented on apps/web/pages/signup.tsx:
```diff
@@ -281,6 +281,7 @@ export default function Signup({
                 username={watch("username")}
                 premium={premiumUsername}
                 usernameTaken={usernameTaken}
+                disabled={!!orgSlug}
```
Reverted to earlier implementation
![image](https://github.com/calcom/cal.com/assets/1780212/f86c4312-dee9-46ac-8608-fd702acc37b7)

hariombalhara committed:
Add test

hariombalhara committed:
Add test

zomars committed:
Merge branch 'main' into 12-12-Disallow_changing_username_and_email_in_case_of_Organization_email_invite

zomars reviewed (dismissed):
Code LGTM. Waiting for checks to pass.

sean-brydon committed:
Merge branch 'main' into 12-12-Disallow_changing_username_and_email_in_case_of_Organization_email_invite

PeerRich committed:
Merge branch 'main' into 12-12-Disallow_changing_username_and_email_in_case_of_Organization_email_invite

sean-brydon committed:
Merge branch 'main' into 12-12-Disallow_changing_username_and_email_in_case_of_Organization_email_invite

hariombalhara committed:
Handle username requirability correctly

hariombalhara committed:
Add more tests

hariombalhara commented:
@zomars There was another related issue that I found after fixing this issue(Test started failing as a side-effect of that). Fixed that as well and updated PR description.

hariombalhara commented on apps/web/pages/signup.tsx:
```diff
@@ -66,16 +68,6 @@ const FEATURES = [
   },
 ];
 
-const getOrgUsernameFromEmail = (email: string, autoAcceptEmailDomain: string) => {
-  const [emailUser, emailDomain = ""] = email.split("@");
-  const username =
-    emailDomain === autoAcceptEmailDomain
-      ? slugify(emailUser)
-      : slugify(`${emailUser}-${emailDomain.split(".")[0]}`);
-
-  return username;
-};
-
```
Moved it to a separate module for reusability across client side and server side.

hariombalhara commented on apps/web/pages/signup.tsx:
```diff
@@ -276,25 +268,29 @@ export default function Signup({
+              {!isOrgInviteByLink ? (
+                <UsernameField
+                  label={t("username")}
+                  username={watch("username") || ""}
```
`username` is optional at client now because it's requirement depends upon if it's an org case or non-org case. It is separately handled now at backend.

hariombalhara commented on apps/web/pages/signup.tsx:
```diff
@@ -276,25 +268,29 @@ export default function Signup({
-                    : `${process.env.NEXT_PUBLIC_WEBSITE_URL}/`
-                }
-              />
+              {!isOrgInviteByLink ? (
```
Don't show the username field if it's invite link. Username would be derived from email automatically. This is same as earlier implementation before regression.

hariombalhara commented on apps/web/pages/signup.tsx:
```diff
@@ -366,7 +362,7 @@ export default function Signup({
                         // If username is present we save it in query params to check for premium
                         const searchQueryParams = new URLSearchParams();
-                        searchQueryParams.set("username", formMethods.getValues("username"));
+                        searchQueryParams.set("username", username);
```
Because username can be nullish, use username variable which is checked for nullishness

hariombalhara commented on apps/web/pages/signup.tsx:
```diff
@@ -402,10 +398,14 @@ export default function Signup({
                         return;
                       }
                       const username = formMethods.getValues("username");
+                      if (!username) {
```
Because there is a theoretical possibility of username being null in the flow. Show the error appropriately if it happens.

hariombalhara commented on apps/web/pages/signup.tsx:
```diff
@@ -491,6 +491,7 @@ export default function Signup({
           </div>
         </div>
       </div>
+      <Toaster position="bottom-right" />
```
Toast needs a toaster which was missing.

hariombalhara committed:
Self review fixes

hariombalhara commented on apps/web/playwright/organization/organization-invitation.e2e.ts:
```diff
@@ -80,10 +87,15 @@ test.describe("Organization", () => {
       await inviteLinkPage.locator("button[type=submit]").click();
       await inviteLinkPage.waitForURL("/getting-started");
+      const dbUser = await prisma.user.findUnique({ where: { email } });
+      expect(dbUser?.username).toBe(usernameDerivedFromEmail);
```
Verify the username

hariombalhara commented on apps/web/playwright/organization/organization-invitation.e2e.ts:
```diff
@@ -96,21 +108,74 @@ test.describe("Organization", () => {
       await page.waitForLoadState("networkidle");
-      await expectInvitationEmailToBeReceived(
+      const inviteLink = await expectInvitationEmailToBeReceived(
         page,
```
Invite link wasn't being tested for in this case. Now it is.

hariombalhara commented on apps/web/playwright/organization/organization-invitation.e2e.ts:
```diff
@@ -38,21 +42,24 @@ test.describe("Organization", () => {
 
-      // eslint-disable-next-line playwright/no-conditional-in-test
-      if (!inviteLink) return null;
+      assertInviteLink(inviteLink);
```
Instead of not running the assertions, fail the test if inviteLink isn't available.

hariombalhara commented on apps/web/playwright/organization/organization-invitation.e2e.ts:
```diff
@@ -96,21 +108,74 @@ test.describe("Organization", () => {
+      await signupPage.close();
+    });
+
+    await test.step("To the organization by invite link", async () => {
```
Test invite link as well for organization email

hariombalhara commented on packages/features/auth/signup/handlers/calcomHandler.ts:
```diff
@@ -52,15 +52,33 @@ async function handler(req: RequestWithUsernameStatus, res: NextApiResponse) {
     foundToken = await findTokenByToken({ token });
     throwIfTokenExpired(foundToken?.expires);
-    await validateUsernameForTeam({ username, email, teamId: foundToken?.teamId ?? null });
+    username = await validateUsernameForTeam({
```
Ensure that the username is the one we obtain from email, even if the user provides a different one in the request.

Udit-takkar committed:
Merge branch 'main' into 12-12-Disallow_changing_username_and_email_in_case_of_Organization_email_invite
